pkg/controller/persistentpodstate/persistent_pod_state_controller.go

```diff
@@ -0,0 +1,451 @@
+				klog.Errorf("parse PersistentPodState(%s/%s) podName(%s) failed: %s",
+					persistentPodState.Namespace, persistentPodState.Name, podName, err.Error())
+				continue
+			}
+			if isInStatefulSetReplicas(index, innerSts) {
+				continue
+			}
+			// others will be deleted for scaling down sts
+			// if pod not exists, then delete PersistentPodState
+			if _, ok := pods[podName]; !ok {
+				delete(newStatus.PodStates, podName)
+			}
+		}
+	}
+	return ctrl.Result{}, r.updatePersistentPodStateStatus(persistentPodState, *newStatus)
+}
+
+func (r *ReconcilePersistentPodState) updatePersistentPodStateStatus(pps *appsv1alpha1.PersistentPodState, newStatus appsv1alpha1.PersistentPodStateStatus) error {
+	if reflect.DeepEqual(pps.Status, newStatus) {
+		return nil
+	}
+	// update PersistentPodState status
+	persistentPodStateClone := &appsv1alpha1.PersistentPodState{}
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		if err := r.Client.Get(context.TODO(), client.ObjectKey{Namespace: pps.Namespace, Name: pps.Name}, persistentPodStateClone); err != nil {
+			return err
+		}
+		persistentPodStateClone.Status = newStatus
+		return r.Client.Status().Update(context.TODO(), persistentPodStateClone)
+	}); err != nil {
+		klog.Errorf("update PersistentPodState(%s/%s) status failed: %s", pps.Namespace, pps.Name, err.Error())
+		return err
+	}
+	klog.V(3).Infof("update PersistentPodState(%s/%s) status from(%s) -> to(%s) success", pps.Namespace, pps.Name, util.DumpJSON(pps.Status), util.DumpJSON(newStatus))
+	return nil
+}
+
+func parseStsPodIndex(podName string) (int, error) {
+	index := strings.LastIndex(podName, "-")
+	return strconv.Atoi(podName[index+1:])
+}
+
+// map[string]*corev1.Pod -> map[Pod.Name]*corev1.Pod
+func (p *ReconcilePersistentPodState) getPodsAndStatefulset(persistentPodState *appsv1alpha1.PersistentPodState) (map[string]*corev1.Pod, *innerStatefulset, error) {
+	inner := &innerStatefulset{}
+	ref := persistentPodState.Spec.TargetReference
+	workload, err := p.finder.GetScaleAndSelectorForRef(ref.APIVersion, ref.Kind, persistentPodState.Namespace, ref.Name, "")
+	if err != nil {
+		klog.Errorf("persistentPodState(%s/%s) fetch statefulSet(%s) failed: %s", persistentPodState.Namespace, persistentPodState.Name, ref.Name, err.Error())
+		return nil, nil, err
+	}
+	inner.Replicas = workload.Scale
+	inner.ReserveOrdinals = workload.ReserveOrdinals
+	inner.DeletionTimestamp = workload.Metadata.DeletionTimestamp
+
+	// DisableDeepCopy:true, indicates must be deep copy before update pod objection
+	pods, _, err := p.finder.GetPodsForRef(ref.APIVersion, ref.Kind, persistentPodState.Namespace, ref.Name, true)
+	if err != nil {
+		klog.Errorf("list persistentPodState(%s/%s) pods failed: %s", persistentPodState.Namespace, persistentPodState.Name, err.Error())
+		return nil, nil, err
+	}
+	matchedPods := make(map[string]*corev1.Pod, len(pods))
+	for i := range pods {
+		pod := pods[i]
+		if pod.DeletionTimestamp.IsZero() {
```

Comment:
Don't have to check deletionTimestamp since you set `active=true` for `p.finder.GetPodsForRef()` above.